### ISSUE TITLE
UI::ProgressBar

### ISSUE THREAD

**Issue by kengey** [SketchUp/api-issue-tracker#360]
Sketchup is lacking a native ProgressBar we developers could use to indicate progress of a long running task (and perhaps even allow the user the cancel a running task). Also, Sketchup on its own could benefit from an indication of progress as well.

I understand this should need some more information and propositions of how this could be implemented. But, to get started, let's have this topic in the issue tracker and collaborate on this.

A few notes already:

1. progress should be tied to an `Operation` (ideally, just like in Revit for example, Model changes that happen outside of an `Operation`, or `Transaction` in the case of Revit, simply throw an error, see https://www.revitapidocs.com/2020/308ebf8d-d96d-4643-cd1d-34fffcea53fd.htm for more info on the Revit `Transaction` class). Maybe it is time for a dedicated `Sketchup::Operation` class.
1. it should be possible for users to cancel a long running task.

Only by taking the 2 points here above into consideration I imagine a `Sketchup::Operation` class that would allow you to start, commit or abort it as well as to cancel (but not `abort` just yet) it, see if it has been cancelled and also to get and set progress.

Since a model can have only 1 operation active, that would create the need for a `Model.active_operation` method, returning the current open operation, if any.

Again, this is not structured, and should more be seen as a vague idea. The TLDR here is that we need a unified and backed by sketchup way to indicate progress to our users.

**Comment by DanRathbun** [comment]
**A very old request that has so far not been implemented !**

There are some interfaces in the API that give clues to possible implementations ...

[`Sketchup/DefinitionList#load_from_url`](http://ruby.sketchup.com/Sketchup/DefinitionList.html#load_from_url-instance_method) allows an abstract `LoadHandler` class. Although the coder had to code the actual progress indicator.

As I recall, Scott L. just wrote a series of pipe symbols to the status bar for the DC extension. But that's not very attractive, and moving the mouse interferes will status bar display.

Note also that the `Sketchup::Http::Request` upload and download progress callbacks are bugged last I knew. (I reported the issue that the values passed into the block are the same each time.)

So yes, either  / or both:

* A class of some kind to do custom progress interfaces, with callback methods (similar to the `LoadHandler`.

* A standard simple messagebox size progress bar that can be updated with the "current percentage" to display (assuming that `1.0` / 100% is the full width of the bar.) The coder determines what work means 100%, and the dialog box would open with 0% initially done. It would be up to the coder to call an update method.
```ruby
pbar = UI::ProgressBar.new("Some Caption")
pbar.message = "Some text to explain what the task is." # appears above the bar.
pbar.on_dismiss("Cancel") do |dialog|
  choice = UI.messagebox("Cancel the current operation? (Y/N)",MB_YESNO)
  if choice != IDYES
    pbar.resume
  else
    # clean up abort etc.
  end
end
pbar.show do |dialog|
  100.times do |i|
    # draw something in the model
    pbar.update(i/100.0)
  end
end
```

Is there any examples from **Shoes** that might give us clues.

**Comment by DanRathbun** [comment]
@kengey You've seen Thomas' example for [TaskBar progress indicator](https://github.com/SketchUp/sketchup-taskbarlist) ? It isn't cross platform, and I wonder what happens if a user has their taskbar hidden ?

**Comment by MSP-Greg** [comment]
I've got a progressbar running in Dialog, but it's tricky.  Originally wrote it for importing drawings from external files.

The main (and SU specific) issue is the fact that adding faces/entities to models slows down as the number of entities in the model increases.  IOW, the time it takes to add 100 faces to a model with 500 faces is much quicker than the time it takes to add 100 faces to a model with 5,000 faces.

Other applications may not have the problem.  Used fibers, and it does slow things down somewhat.

But I'd agree that something native would be helpful.  For long operations, users needs something showing that progress is being made, as opposed to 'is this frozen?'...

**Comment by kengey** [comment]
@DanRathbun There are all sorts of ways to do this ourselves with native code (and thus native threads) and a cross platform gui platform, such as gtk (https://www.gtk.org/). But, this should be a uniform something available in the api. I once made one in 4d Virtual Builderr, somewhere in 2012-2013 (https://extensions.sketchup.com/en/content/4d-virtual-builder-trimble-sketchup). You can somewhat see it in action here: https://youtu.be/8tJrSxPKOC4?t=2646 although it looks as if it is part of the background window, it is a simple viewport overlay covering the entire sketchup viewport.

Another possibility is to use a HtmlDialog. But, the issue is that it seems to use some thread syncing, which means that a blocking action in ruby also blocks the HtmlDialog from updating. That would be solved if:
1. sketchup solves #347 
1. we split the work in queable workitems and perform a work-item per timer cycle.... I've done some internal proof of concepts doing just that, but it significally takes more time to perform a task. And, it does not block the ui, so users can screw things up. Ideally, UI still freezes, but the user recieves feedback.

**Comment by DanRathbun** [comment]
Yes I remember all attempts at using WebDialog and HtmlDialog were "less than ideal".

**Comment by DanRathbun** [comment]
> For long operations, users needs something showing that progress is being made, as opposed to 'is this frozen?'...

Win Common Dialogs have a marque style for progress bars (like the ol' "KITT" Firebird scanlights or "Cylon" eyes. Or remember the lights below the viewscreen on Star Trek?)

I wonder if Carbon or Cocoa has similar for the Mac platform ?

Is we could also have a **`UI::BusyBar`** class.

**Comment by DanRathbun** [comment]
> 1. progress should be tied to an `Operation`

I've been brainstormin' a boilerplate and realized that this idea would limit some advanced workflows that work with multiple open models on the Mac, or successive open models on Win. In this scenario it would be the coder's responsibility to wrap the multiple model operations properly.

Now I don't mean to say that a `UI::ProgressOperation` class can't be implemented for simpler tasks on a single model. It could be beneficial for novice coders.

**Comment by kengey** [comment]
> this idea would limit some advanced workflows

I agree.

**Comment by DanRathbun** [comment]
> Is there any examples from **Shoes** that might give us clues. (_To a Mac implementation_.)

I looked at Shoes but it has so much code wrapped around everything I could find the "real meat".

However, I did find an open source example in a blog ...
https://blog.eternalstorms.at/2015/05/06/how-to-animate-a-determinate-nsprogressindicator-open-source/

He posted his code on GitHub ...
https://github.com/eternalstorms/NSProgressIndicator-ESSProgressIndicatorCategory

He also has posted a busy indicator ...
https://github.com/eternalstorms/ESSSquareProgressIndicator

**Comment by DanRathbun** [comment]
I updated the example boilerplate in the 2nd post.

I'd be interesting to know if coders prefer a class like shown (above, pots 2) where blocks are defined for events upon an instance, or if coders prefer abstract classes where the entire class with callback methods are defined (instead of blocks.)

**Comment by kengey** [comment]
I think this progressbar should be a global, like `Sketchup.set_status_text(status_text = '', position = SB_PROMPT) ⇒ nil`, thus something like `Sketchup.set_progress(double)`. The reason is I believe there should be just one progress bar, being part of the ui. Maybe even only a tool would be allowed to set progress, since there can only be one tool active.

These are just thoughts.

**Comment by thomthom** [comment]
Closing this one in favor of the existing #360. (They'll be referenced anyway.)

**Comment by DanRathbun** [comment]
@thomthom UH ... this **IS** 360 !

**Comment by thomthom** [comment]
Sorry, mean #220. Copypasta error.

**Comment by DanRathbun** [comment]
**Again**, this issue is the one we are participating in. REOPEN it and **close the other**.

**Comment by DanRathbun** [comment]
Thank you!

**Comment by Matherone** [comment]
Yes please. 

Regarding the progress bar UI element, ~~it would be nifty, until the progress bar is live, to have a progress bar that tracks the progress of the development of the progress bar?~~ could I humbly suggest adding support for an accompanying text string above the progress bar, where developers may serve random one-line jokes for the disgruntled, or daily aphorisms for the more philosophically inclined, to further alleviate the plight of the waiting end users?

For instance:

»Good things come to those who wait«
»Better late than never«
»Fools rush in«

![Still waiting for Sketchup  Surely a response will come (meme, skeleton, computer, fun, humor)](https://user-images.githubusercontent.com/44017984/89877739-d01d3400-dbc0-11ea-8b25-211e54e5bb9a.jpg)

– 🙂

**Comment by DanRathbun** [comment]
As I wrote in the boilerplate example, we indeed should have message text above the bar ...
```ruby
pbar.message= "Some text to explain what the task is." # appears above the bar.
```

----

In addtion ...

A `UI::ProgressBar` object should also be instantiatable using **hash** / **named properties**, ie:
```ruby
progress = UI::ProgressBar.new(
  caption: "Some Caption",
  message: "Some text to explain what the task is.",
  estimation_prompt: "Estimated time remaining: %d min %f secs.",
  estimate: [2, 0.0],
  range: 1000
)
```
... or a **block** pattern (similar to a gemspec file) ...
```ruby
progress = UI::ProgressBar.new do |pbar|
  pbar.caption= "Some Caption"
  pbar.message= "Some text to explain what the task is."
  pbar.estimation("Estimated time remaining: %d min %f secs.")
  pbar.estimate(2, 0.0)
  pbar.range= 1000
end
```

**Comment by QROST** [comment]
Any updates on this issue?
